### PR TITLE
Feature falcor objects

### PR DESCRIPTION
**Pull request by gyllen**
Updated pull request

### CODE REVIEW

File: src/component/DetectionComponent.ts [hunk 1]
```diff
@@ -1,127 +1,109 @@
...
 
 import {Node} from "../Graph";
 import {Container, Navigator} from "../Viewer";
-import {APIv2, IRect} from "../API";
+import {APIv3} from "../API";
 
 import {ComponentService, Component} from "../Component";
+import {IVNodeHash} from "../Render";
+
+interface IRect {
```
Comment (oscarlorentzon) [inline src/component/DetectionComponent.ts]: Is this a general interface? Should it be moved to the api/interfaces folder?

Comment (oscarlorentzon) [inline src/component/DetectionComponent.ts]: This interface contains more information than just a rectangle. IDetection feels like a more informative name.

File: src/component/DetectionComponent.ts [hunk 2]
```diff
@@ -1,127 +1,109 @@
...
-                this.setRectContainer(node.image.width, node.image.height);
-                this.removeRectsFromDOM();
-
-                this.apiV2.im.callOr(node.key).then((data: any) => {
-                    this.detectionData = data;
-                    this.updateRects(this.detectionData.or_rects);
-                });
-            });
+        this._disposable = this._navigator.stateService.currentNode$.flatMap((node: Node): rx.Observable<any> => {
+            let path: string = "['imageByKey']";
```
Comment (oscarlorentzon) [inline src/component/DetectionComponent.ts]: Can this string creation be encapsulated within the APIv3 class?

File: src/component/DetectionComponent.ts [hunk 3]
```diff
@@ -1,127 +1,109 @@
...
+            path += "['" + node.key + "']";
+            path += "['ors'][0..20]['key', 'obj', 'rect', 'value', 'package', 'score']";
+
+            return this.apiV3.model.get(path);
+        }).map((ors: any): IVNodeHash => {
+            let rects: IRect[] = [];
+            delete ors.json.imageByKey.$__path;
```
Comment (oscarlorentzon) [inline src/component/DetectionComponent.ts]: Enforcing the user of the Falcor API to delete this property every time seems like a possible cause of future bugs. Can this be encapsulated in the APIv3 in any way?

File: src/component/DetectionComponent.ts [hunk 4]
```diff
@@ -1,127 +1,109 @@
...
+            for (let or in ors) {
+                if (ors.hasOwnProperty(or)) {
+                    or = ors[or];
+
+                    if (!or) {
+                        continue;
+                    }
+
+                    let r: number[] = [];
+                    r[0] = or.rect.geometry.coordinates[1][0];
```
Comment (oscarlorentzon) [inline src/component/DetectionComponent.ts]: This is dynamic. Could the API expose interfaces for this structure to make it typed?

File: src/component/marker/MarkerComponent.ts [hunk 5]
```diff
@@ -107,12 +108,33 @@ export class MarkerComponent extends Component {
                 return this.renderHash(args);
             })
             .subscribe(this._container.glRenderer.render$);
+
+        this._disposableMapillaryObject = null;
+        this.configuration$.subscribe((conf: IMarkerConfiguration) => {
```
Comment (oscarlorentzon) [inline src/component/marker/MarkerComponent.ts]: The external state (`_disposableMapillaryObject`) can be avoided by doing (pseudo):

`this._configurationSubscription = this.configuration$.
    filter(conf.mapillaryObject)
    flatMapLatest<Marker>(return this._navigator.vectorTilesService.mapillaryObjects$.map<Marker>(create marker))
    subscribe(do things with marker)`

File: src/geo/GoogleTiles.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import {IGoogleTile, IPoint, IBBox} from "../Geo";
+import {ILatLon} from "../Graph";
+
+export class GoogleTiles {
```
Comment (oscarlorentzon) [inline src/geo/GoogleTiles.ts]: This class has some magic numbers that would be easier to understand if they were made constants.
The class can also easily be unit tested to provide documentation of how it works.

File: src/graph/VectorTilesService.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            let z: number = parseInt(Object.keys(tile.json.tile.all)[0], 10);
+            let x: number = parseInt(Object.keys(tile.json.tile.all[z])[0], 10);
+            let y: number = parseInt(Object.keys(tile.json.tile.all[z][x])[0], 10);
+            let objects: any = tile.json.tile.all[z][x][y].objects;
+
+            let ret: MapillaryObject[] = [];
+
+            for (let object in objects) {
+                if (objects.hasOwnProperty(object)) {
+                    if (object !== "$__path") {
```
Comment (oscarlorentzon) [inline src/graph/VectorTilesService.ts]: The $__path property is handled differently here and in the DetectionComponent.
Also indentation can be decreased by using `!objects.hasOwnProperty(object)` and `continue`.

File: src/viewer/interfaces/IViewerOptions.ts [hunk 8]
```diff
@@ -38,6 +38,12 @@ export interface IViewerOptions {
     direction?: boolean;
 
     /**
+     * Show detection rectangles in images
+     * @default true
```
Comment (oscarlorentzon) [inline src/viewer/interfaces/IViewerOptions.ts]: Docs default value does not correspond to behavior in ComponentController.

File: styles/mapillary-js.css [hunk 9]
```diff
@@ -29,7 +29,7 @@
     right: 0;
     bottom: 0;
     left: 0;
-    z-index: 10;
+    z-index: -1;
```
Comment (oscarlorentzon) [inline styles/mapillary-js.css]: Feels like the DetectionComponent styles should be in a DetectionComponent.css file and not in the general mapillary-js css file.

**Comment by oscarlorentzon** [comment]
The api/interfaces/IRect.ts interface is unused after the changes in this PR. If the Falcor API should be used for retrieving rects then that interface should be removed.

File: src/component/DetectionComponent.ts (hunk 1 above)
Comment (gyllen) [inline src/component/DetectionComponent.ts]: Changed name, Im not sure I want to move interfaces to APIv3 repo yet. Not 100% sure on interrface.

File: src/component/DetectionComponent.ts (hunk 2 above)
Comment (gyllen) [inline src/component/DetectionComponent.ts]: I updated this to use the array API of falcor. Better than building the string.

File: src/component/DetectionComponent.ts (hunk 3 above)
Comment (gyllen) [inline src/component/DetectionComponent.ts]: Yeah this is an issue, soon with a solution though, tracked here -> https://github.com/Netflix/falcor/issues/708

File: src/component/DetectionComponent.ts (hunk 4 above)
Comment (gyllen) [inline src/component/DetectionComponent.ts]: I would need to think a little about how to type falcor response. No good idea right now.

File: src/viewer/interfaces/IViewerOptions.ts (hunk 8 above)
Comment (gyllen) [inline src/viewer/interfaces/IViewerOptions.ts]: Fixed

File: styles/mapillary-js.css (hunk 9 above)
Comment (gyllen) [inline styles/mapillary-js.css]: Fixed

**Comment by gyllen** [comment]
Removed IRect.ts since no longer needed

File: src/geo/GoogleTiles.ts (hunk 6 above)
Comment (gyllen) [inline src/geo/GoogleTiles.ts]: Class is just a translation of standard javascript for lon, lat translation to google tiles. Added reference to stackoverflow post.

File: src/graph/VectorTilesService.ts (hunk 7 above)
Comment (gyllen) [inline src/graph/VectorTilesService.ts]: Fixed

**Pull request merged**